### PR TITLE
Accept c_ptr(c_char) in string/bytes factory functions

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR adds the ability to create `string` and `bytes` from `c_char` pointers.

Resolves https://github.com/chapel-lang/chapel/issues/16321

Summary of changes:
- Change the argument type of factories that creates the type from
  `c_ptr(uint(8))` to `c_ptr(?t)`
- Give a compiler error if `t` is not `uint(8)` or `c_char`
- Cast the received argument to `bufferType` (`c_ptr(uint(8))`) before calling
  internal helpers
- Modify tests to pass `c_ptr(c_char)` to those functions
- Add tests for the error message for wrong types
- Adjust documentation

Test:
- [x] make docs
- [x] standard

### CODE REVIEW

File: modules/internal/Bytes.chpl [hunk 1]
```diff
@@ -174,21 +174,24 @@ module Bytes {
...
   /*
      Creates a new :mod:`bytes <Bytes>` which borrows the memory allocated for a
-     `c_ptr(uint(8))`. If the buffer is freed before the :mod:`bytes <Bytes>`
-     returned from this function, accessing it is undefined behavior.
+     `c_ptr`. If the buffer is freed before the :mod:`bytes <Bytes>` returned
+     from this function, accessing it is undefined behavior.
 
      :arg s: Buffer to borrow
-     :type s: `bufferType` (i.e. `c_ptr(uint(8))`)
+     :type x: `c_ptr` to `uint(8)` or `c_char`
```
Comment (ben-albrecht) [inline modules/internal/Bytes.chpl:181]: Would you be open to writing both types out?

```suggestion
     :type x: `c_ptr(uint(8))` or `c_ptr(c_char)`
```

Comment (ben-albrecht) [reply modules/internal/Bytes.chpl:181]: rationale: I think it's easier read quickly and leaves less up for interpretation.

This applies in other docs as well.

File: modules/internal/Bytes.chpl [hunk 2]
```diff
@@ -174,21 +174,24 @@ module Bytes {
...
      :returns: A new :mod:`bytes <Bytes>`
   */
-  inline proc createBytesWithBorrowedBuffer(x: bufferType, length: int, size: int) {
+  inline proc createBytesWithBorrowedBuffer(x: c_ptr(?t), length: int, size: int) {
+    if t != byteType && t != c_char {
+      compilerError("Cannot create a bytes with a buffer of ", t:string);
+    }
     var ret: bytes;
-    initWithBorrowedBuffer(ret, x, length,size);
+    initWithBorrowedBuffer(ret, x:bufferType, length,size);
```
Comment (ben-albrecht) [inline modules/internal/Bytes.chpl:194]: nit:
```suggestion
    initWithBorrowedBuffer(ret, x:bufferType, length, size);
```

**Review by ben-albrecht — APPROVED** [review]
Changes look good. See question about docs inline.

File: modules/internal/Bytes.chpl (hunk 1 above)
Comment (e-kayrakli) [reply modules/internal/Bytes.chpl:181]: Yeah, that makes sense. I'll make that change

**Pull request merged**